executor/include/executor/default/base_executor.hpp

```diff
@@ -14,46 +14,4 @@
 #include <array>
 #include <executor/trait/is_executor_available.hpp>
 #include <executor/trait/is_instance_of_base.hpp>
-#include <functional>
-#include <memory>
-#include <stdexcept>
 #include <string>
-
-namespace executor {
-
-template <template <typename...> class Derived, typename Blocking,
-          typename ProtoAllocator>
-struct base_executor {
-  template <typename Executor,
-            typename executor::instance_of_base<Derived, Executor> = 0>
-  bool operator==(const Executor& rhs) const noexcept {
-    return std::is_same<Derived<Blocking, ProtoAllocator>, Executor>::value;
-  }
-
-  template <typename Executor,
-            typename executor::instance_of_base<Derived, Executor> = 0>
-  bool operator!=(const Executor& rhs) const noexcept {
-    return !operator==(rhs);
-  }
-
-  static constexpr bool query(const blocking_t::always_t& t) noexcept {
-    return std::is_same<Blocking, blocking_t::always_t>();
-  }
-
-  static constexpr bool query(const blocking_t::never_t& t) noexcept {
-    return std::is_same<Blocking, blocking_t::never_t>();
-  }
-
-  static constexpr bool query(const blocking_t::possibly_t& t) noexcept {
-    return std::is_same<Blocking, blocking_t::possibly_t>();
-  }
-
-  template <typename F, typename... Args>
-  void bulk_execute(F&& f, Args&&... args, std::size_t n) const {
-    for (std::size_t i = 0; i < n; ++i) {
-      std::forward<F>(f)(std::forward<Args>(args)..., i);
-    }
-  }
-};
```

Comment:
Given your answer in https://github.com/shrijitsingh99/unified-executors/pull/12#discussion_r450110138 I need to revise my comments here. You mentioned

> Also, regarding `base_executor`there is not needed and I have removed it in an upcoming PR. There is no requirement that the executors satisfy all those 4 properties. An executor can implement any set of properties.
> Adding on that out of the 4 to have been removed as they are no longer part of the proposal namely, Interface and Cardinality.

I still think it's advantageous to have a base class so that you don't have to rewrite the same query functions everytme. However, your base class design might need to encompass this nature that an executor is free to implement whatever properties it desires. The new information tells, me that you're executors should at least follow a mix-in pattern. I'm not saying a dedicated common base class needs to exists for it, but every property your class implements, should expose the methods it requires.

Give me some time to exemplify.

Reply:
Here https://godbolt.org/z/9kGd7q

Reply:
Beautiful :)

Riposte: https://godbolt.org/z/xeyNUM

Reply:
This design looks very promising and is quite clean too.
The reason I am holding out is this design makes the assumption that the properties do not have state information so having a `query()` in `constexpr_executor_property` would not allow such stateful properties.

Though this design certainly removes the need for having `require` and `query` in the class, which is super.

I think we should first have a discussion on what all properties we will need for PCL whether this design matches the proposal (if we even want to ensure compatibility). Then move forward with this design.

Reply:
> this design makes the assumption that the properties do not have state information so having a `query()` in `constexpr_executor_property`

You're right, but it can be... remedied. https://godbolt.org/z/zhKM6e

> I think we should first have a discussion on what all properties we will need for PCL

Sure